src/main/java/com/cleanroommc/groovyscript/compat/mods/jei/Ingredient.java

```diff
@@ -0,0 +1,252 @@
+     */
+    @GroovyBlacklist
+    public void applyChanges(IIngredientRegistry ingredientRegistry) {
+        if (hideAllIngredients) {
+            ingredientRegistry.getRegisteredIngredientTypes().forEach(hiddenTypes::add);
+        }
+        for (var type : hiddenTypes) {
+            hiddenIngredients.computeIfAbsent(type, k -> new ArrayList<>()).addAll(ingredientRegistry.getAllIngredients(type));
+        }
+        for (var entry : hiddenIngredients.entrySet()) {
+            if (entry.getValue().isEmpty()) continue;
+            //noinspection unchecked,rawtypes
+            ingredientRegistry.removeIngredientsAtRuntime(entry.getKey(), ingredientRegistry.getIngredientHelper(entry.getKey()).expandSubtypes((List) entry.getValue()));
+        }
+
+        for (var entry : addedIngredients.entrySet()) {
+            //noinspection unchecked,rawtypes
+            ingredientRegistry.addIngredientsAtRuntime(entry.getKey(), ingredientRegistry.getIngredientHelper(entry.getKey()).expandSubtypes((List) entry.getValue()));
+        }
+
+    }
+
+    @Override
+    public void onReload() {
+        addedIngredients.clear();
+        hiddenIngredients.clear();
+        hiddenTypes.clear();
+        hideAllIngredients = false;
+    }
+
+    @MethodDescription(
+            type = MethodDescription.Type.ADDITION,
+            example = @Example(value = "VanillaTypes.ITEM, item('minecraft:bed').withNbt([display:[Name:'Beds come in 16 different colors!']])", imports = "mezz.jei.api.ingredients.VanillaTypes")
+    )
+    public void add(IIngredientType<?> type, Collection<Object> entries) {
+        addedIngredients.computeIfAbsent(type, k -> new ArrayList<>()).addAll(entries);
+    }
+
+    @MethodDescription(type = MethodDescription.Type.ADDITION)
+    public void add(IIngredientType<?> type, Object... entries) {
+        add(type, Arrays.asList(entries));
+    }
+
+    @MethodDescription(type = MethodDescription.Type.ADDITION, example = @Example("item('minecraft:stone:1').withNbt([display:[Name:'Special Granite']])"))
+    public void add(IIngredient ingredient) {
+        if (IngredientHelper.isEmpty(ingredient)) {
+            GroovyLog.msg("Error adding items {}", ingredient)
+                    .add("Items must not be empty")
+                    .error()
+                    .post();
+            return;
+        }
+        if (IngredientHelper.isItem(ingredient)) {
+            addedIngredients.computeIfAbsent(VanillaTypes.ITEM, k -> new ArrayList<>()).addAll(Arrays.asList(ingredient.getMatchingStacks()));
+        } else if (IngredientHelper.isFluid(ingredient)) {
+            addedIngredients.computeIfAbsent(VanillaTypes.FLUID, k -> new ArrayList<>()).add(IngredientHelper.toFluidStack(ingredient));
+        } else {
+            GroovyLog.msg("Error adding items {}", ingredient)
+                    .add("Could not identify the type")
+                    .error()
+                    .post();
+        }
```

Comment:
You can get the ingredient type with `mezz.jei.api.ingredients.IIngredientRegistry.getIngredientType()`

Reply:
we dont have access to the `IIngredientRegistry` yet when we want to run this. as an example, trying this code snippet in place of the if/else block generates an NPE
```java
        hiddenIngredients.computeIfAbsent(JeiPlugin.itemRegistry.getIngredientType(ingredient), k -> new ArrayList<>()).addAll(Arrays.asList(ingredient.getMatchingStacks()));
```